// minecraft-java-core/build/Authenticator/Mojang.js
"use strict";
/**
 * @author Luuxis
 * @license CC-BY-NC 4.0 - https://creativecommons.org/licenses/by-nc/4.0/
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChangeAuthApi = exports.signout = exports.validate = exports.refresh = exports.login = void 0;
const crypto_1 = __importDefault(require("crypto"));
const node_fetch_1 = __importDefault(require("node-fetch"));
let api_url = 'https://authserver.mojang.com';
let auth;
function login(username, password) {
    let UUID = crypto_1.default.randomBytes(16).toString('hex');
    new Promise(async (resolve, reject) => {
        if (!password) {
            auth = {
                access_token: UUID,
                client_token: UUID,
                uuid: UUID,
                name: username,
                user_properties: '{}',
                meta: {
                    offline: true,
                    type: 'Mojang'
                }
            };
            return resolve(auth);
        }
        let post = {
            agent: {
                name: "Minecraft",
                version: 1
            },
            username,
            password,
            clientToken: UUID,
            requestUser: true
        };
        let message = await (0, node_fetch_1.default)(`${api_url}/authenticate`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(post)
        }).then(res => res.json());
        if (message.error) {
            auth = message;
            return reject(message);
        }
        ;
        let user = {
            access_token: message.accessToken,
            client_token: message.clientToken,
            uuid: message.selectedProfile.id,
            name: message.selectedProfile.name,
            user_properties: '{}',
            meta: {
                offline: false,
                type: 'Mojang'
            }
        };
        auth = user;
        return resolve(user);
    });
    return auth;
}
exports.login = login;
async function refresh(acc) {
    new Promise(async (resolve, reject) => {
        let post = {
            accessToken: acc.access_token,
            clientToken: acc.client_token,
            requestUser: true
        };
        let message = await (0, node_fetch_1.default)(`${api_url}/refresh`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(post)
        }).then(res => res.json());
        if (message.error) {
            auth = message;
            return reject(message);
        }
        ;
        let user = {
            access_token: message.accessToken,
            client_token: message.clientToken,
            uuid: message.selectedProfile.id,
            name: message.selectedProfile.name,
            user_properties: '{}',
            meta: {
                offline: false,
                type: 'Mojang'
            }
        };
        auth = user;
        return resolve(user);
    });
    return auth;
}
exports.refresh = refresh;
async function validate(acc) {
    new Promise(async (resolve, reject) => {
        let post = {
            accessToken: acc.access_token,
            clientToken: acc.client_token,
        };
        let message = await (0, node_fetch_1.default)(`${api_url}/validate`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(post)
        });
        if (message.status == 204) {
            auth = true;
            return resolve(true);
        }
        else {
            auth = false;
            return reject(false);
        }
    });
    return auth;
}
exports.validate = validate;
async function signout(acc) {
    new Promise(async (resolve, reject) => {
        let post = {
            accessToken: acc.access_token,
            clientToken: acc.client_token,
        };
        let message = await (0, node_fetch_1.default)(`${api_url}/invalidate`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(post)
        }).then(res => res.text());
        if (message == "") {
            auth = true;
            return resolve(true);
        }
        else {
            auth = false;
            return reject(false);
        }
    });
    return auth;
}
exports.signout = signout;
function ChangeAuthApi(url) {
    api_url = url;
}
exports.ChangeAuthApi = ChangeAuthApi;
